### PR TITLE
Make NestedObjectMapper its own class

### PR DESCRIPTION
**Pull request by romseygeek**
Nested objects are implemented via a Nested class directly on object mappers,
even though nested and non-nested objects have quite different semantics. In
addition, most call-sites that need to get an object mapper in fact need a nested
object mapper. To make it clearer that nested and object mappers are different
beasts with different implementations and different requirements, we should
split them into different classes.

### CODE REVIEW

**Comment by romseygeek** [comment]
Take 2 of this, starting again from #73058 but with a straight copy of the ObjectMapper merging code and more tests.

**Comment by romseygeek** [comment]
I've merged in the tests from #74899 here as well.  Unfortunately it looks like we need to keep `fixRedundantIncludes` as a method on RootObjectMapper, rather than having it contained entirely with the NestedObjectMapper builder, mainly due to the way that index templates are merged (because merges are done on realised Mappers and we don't go back via Builders)

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 1]
```diff
@@ -160,20 +81,17 @@ public Builder dynamic(Dynamic dynamic) {
...
-
         public Builder add(Mapper.Builder builder) {
             mappersBuilders.add(builder);
             return this;
         }
 
-        @Override
-        public ObjectMapper build(ContentPath contentPath) {
-            contentPath.add(name);
+        public List<Mapper.Builder> builders() {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:89]: These look unused, maybe we can remove?

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 2]
```diff
@@ -160,20 +81,17 @@ public Builder dynamic(Dynamic dynamic) {
...
+            return mappersBuilders;
+        }
 
+        protected Map<String, Mapper> buildMappers(ContentPath contentPath) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:93]: Small comment, would be nice to make this `protected final` to clarify it's not meant to be overridden.

File: server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+ */
+public class NestedObjectMapper extends ObjectMapper {
+
+    public static final String CONTENT_TYPE = "nested";
+
+    public static class Builder extends ObjectMapper.Builder {
+
+        private Explicit<Boolean> includeInRoot = new Explicit<>(false, false);
+        private Explicit<Boolean> includeInParent = new Explicit<>(false, false);
+        private boolean redundantIncludesFixed = false;
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java:34]: This variable looks unused.

**Review by jtibshirani — COMMENTED** [review]
This makes sense to me, I just left small comments. Could you briefly explain the bug that required us to revert the first version of this PR, and how we avoid it now? I'm not totally clear on what the problem was.

**Comment by romseygeek** [comment]
> Could you briefly explain the bug that required us to revert the first version of this PR, and how we avoid it now?

The `enabled` parameter wasn't being preserved during merges, which meant that if you tried to put the same mapping twice with an explicit `enabled` parameter you'd get merge conflicts.  There's now an explicit test (`testEnabled()`) in NestedObjectMapperTests for it.  The first iteration of this code had a completely rewritten `merge` method, but I've gone back to sharing most of the code with ObjectMapper.  In a follow-up I want to change this implementation so that NestedObjectMapper doesn't actually extend ObjectMapper at all, at which point I'll need to disentangle everything again, but we have better test coverage so I'm happy to leave it for now and keep this PR relatively simple.

**Review by jtibshirani — APPROVED** [review]
Thanks for explaining!

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**